crates/pessimistic-proof/src/local_state.rs

```diff
@@ -16,38 +19,141 @@ pub struct LocalNetworkState {
     /// Commitment to the [`BridgeExit`].
     pub exit_tree: LocalExitTree<Keccak256Hasher>,
     /// Commitment to the balance for each token.
-    pub balance_tree: BalanceTree,
+    pub balance_tree: LocalBalanceTree<Keccak256Hasher>,
+    /// Commitment to the Nullifier Set for the local network, tracks claimed assets on foreign networks
+    pub nullifier_set: NullifierTree<Keccak256Hasher>,
 }
 
 impl LocalNetworkState {
-    /// Apply the [`Certificate`] on the current [`State`].
+    pub fn roots(&self) -> LeafProofOutput {
+        (self.exit_tree.get_root(), self.balance_tree.root, self.nullifier_set.root)
+    }
+
+    /// Apply the [`BatchHeader`] on the current [`State`].
     /// Returns the commitment on the resulting state if successful.
-    pub fn apply_certificate(
+    pub fn apply_batch_header(
         &mut self,
-        certificate: &Certificate,
-    ) -> Result<(ExitRoot, BalanceRoot), ProofError> {
+        batch_header: &BatchHeader<Keccak256Hasher>,
+    ) -> Result<(), ProofError> {
         // Check the initial state
         let computed_root = self.exit_tree.get_root();
-        if computed_root != certificate.prev_local_exit_root {
-            return Err(ProofError::InvalidLocalExitRoot {
+        if computed_root != batch_header.prev_local_exit_root {
+            return Err(ProofError::InvalidInitialLocalExitRoot {
                 got: computed_root,
-                expected: certificate.prev_local_exit_root,
+                expected: batch_header.prev_local_exit_root,
             });
         }
+        if self.balance_tree.root != batch_header.prev_balance_root {
+            return Err(ProofError::InvalidInitialBalanceRoot);
+        }
+        if self.nullifier_set.root != batch_header.prev_nullifier_root {
+            return Err(ProofError::InvalidInitialNullifierRoot);
+        }
+
+        // TODO: benchmark if BTreeMap is the best choice in terms of SP1 cycles
+        let mut new_balances: BTreeMap<_, _> =
+            batch_header.balances_proofs.iter().map(|(k, v)| (*k, v.0)).collect();
+
+        // TODO: check batch_header.imported_exits_root
+        // Apply the imported bridge exits
+        for (imported_bridge_exit, nullifier_path) in &batch_header.imported_bridge_exits {
+            if matches!(imported_bridge_exit.bridge_exit.leaf_type, LeafType::Message) {
+                // TODO: handle bridge messages
+                panic!()
```

Comment:
Can you expand on that? I didn't really find docs on bridge messages. How do they differ from bridge transfers?

Reply:
In terms of doc, maybe these slides from slide 5: https://github.com/0xPolygonHermez/zkevm-techdocs/blob/main/slides/zkevm-architecture-part3-layer-communication.pdf

When `LeafType` is `Message`, the `metadata` field carries the payload for the function call to the contract `bridge_exit.destination_address`. My understanding is that we do not need to care about the semantic of this contract call in the pessimistic proof, so in that sense I believe that it could take the same execution path as for the native token asset transfers as in skipping the insertion in the balance tree (but still being included in the local exit tree)

That said I'm ok having that in another PR to check all that properly :ok_hand:

Reply:
Paging @invocamanman - I **think** messages can have native ETH transfers so it could be a bit more complicated to parse that part of the message and update balances. 🙃

Reply:
Thanks for the details. I think it's best to leave this to a separate PR. 
One thought though, maybe it the API would be a bit cleaner if `BridgeExit` was an enum with two variants (`Transfer` and `Message`) instead of the current version where `BridgeExit` holds an enum `leaf_type: LeafType`.